pallets/author-filter/src/lib.rs

```diff
@@ -0,0 +1,126 @@
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Moonbeam.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Small pallet responsible determining which accounts are eligible to author at the current
+//! block height.
+//!
+//! Currently this pallet is tightly coupled to our stake pallet, but this design
+//! should be generalized in the future.
+//!
+//! Using a randomness beacon supplied by the `Randomness` trait, this pallet takes the set of
+//! currently staked accounts from pallet stake, and filters them down to a pseudorandom subset.
+//! The current technique gives no preference to any particular author. In the future, we could
+//! disfavor authors who are authoring a disproportionate amount of the time in an attempt to
+//! "even the playing field".
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use frame_support::pallet;
+
+pub use pallet::*;
+
+#[pallet]
+pub mod pallet {
+
+	use frame_support::pallet_prelude::*;
+	use frame_support::traits::Randomness;
+	use frame_support::traits::Vec;
+	use frame_system::pallet_prelude::*;
+	use sp_core::H256;
+	use sp_runtime::Percent;
+
+	/// The Author Filter pallet
+	#[pallet::pallet]
+	pub struct Pallet<T>(PhantomData<T>);
+
+	/// Configuration trait of this pallet.
+	#[pallet::config]
+	pub trait Config: frame_system::Config + stake::Config {
+		/// The overarching event type
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+		/// Deterministic on-chain pseudo-randomness used to do the filtering
+		type RandomnessSource: Randomness<H256>;
+	}
+
+	impl<T: Config> author_inherent::CanAuthor<T::AccountId> for Pallet<T> {
+		fn can_author(account: &T::AccountId) -> bool {
+			CurrentEligible::<T>::get().contains(account)
+		}
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		// At the beginning of each block, we calculate the set of eligible authors for this block.
+		// TODO Design Decision:
+		// If we move this logic to on_finalize to calculate for the next block, we get to know in
+		// advance who the next eligible authors are. That is nice because it is easy to know in
+		// from offchain who will author next. You just need to read storage.
+		// On the other hand, it leads to liveness attacks. If the eligible authors collude to not
+		// author, then the chain is bricked. We can 't even force them out with governance because
+		// governance stops when the chain is stalled. In that way, the `EligibleRatio` _is_ our
+		// security assumption. By leaving this in on_initialize, we can rely on Polkadot's
+		// randomness beacon having a different value when there is a different relay parent.
+		fn on_initialize(_: T::BlockNumber) -> Weight {
+			let mut staked: Vec<T::AccountId> = stake::Module::<T>::validators();
+
+			let num_eligible = EligibleRatio::<T>::get() * staked.len();
```

Comment:
I wonder if we want a percentage, or if we want a minimal amount. Cause if we have 16 collators, selecting ~8 is still a lot no ?

Reply:
Actually, could we make it a storage item so we could modify it without upgrade.

Reply:
Good idea. It's a storage item so we can change it with governance (sudo or democracy)